File: Runtime/TypeScript/KFData/Format/KFDName.ts

/*
*
 *给所有的字符串分配一个ID吧
  * */

export class KFDNameStrings
{
    public  constructor()
    {
        this._ID2Strings.push("");
    }

    public GetNameID(namestr:string):number
    {
        if(namestr == "")
            return 0;
        let nameid = this._Strings2ID[namestr];
        if(nameid)
        {
            return nameid
        }
        return this.__NewNameID(namestr);
    }

    public GetNameStr(nameid:number):string
    {
        if(nameid <= 0)
            return "";
        return this._ID2Strings[nameid];
    }

    private __NewNameID(namestr:string):number
    {
        this._ID2Strings.push(namestr);
        ///ID=数组的长度-1
        let nameid:number = this._ID2Strings.length - 1;
        this._Strings2ID[namestr] = nameid;
        return nameid;
    }

    ///找字符串的索引
    private _Strings2ID:{[key: string]: number;} = {};

    ///所有字符串的集合
    private _ID2Strings:Array<string> = new Array<string>();
}

export class KFDName
{
    public constructor(namestr:string = "")
    {
        this.value = KFDName._Strs.GetNameID(namestr);
    }

    public setString(namestr:string)
    {
        this.value = KFDName._Strs.GetNameID(namestr);
    }

    public toString():string
    {
        return KFDName._Strs.GetNameStr(this.value);
    }

    public value:number = 0;
    public static _Strs:KFDNameStrings = new KFDNameStrings();
}

File: Runtime/CPP/KFData/Format/KFDName.cpp
#include "KFDName.h"

NS_KF_BEGIN

KFDNameStrings::KFDNameStrings()
{
	_ID2Strings.push_back("");
}

int32 KFDNameStrings::GetNameID(const kfstr& namestr)
{
	if(namestr.empty()) return 0;

	auto index = _Strings2ID.find(namestr);
	if (index == _Strings2ID.end())
	{
		return __NewNameID(namestr);
	}

	return index->second;
}

int32 KFDNameStrings::__NewNameID(const kfstr& namestr)
{
	_ID2Strings.push_back(namestr);
	///ID=数组的长度-1
	int32 nameid = kfSize2I(_ID2Strings.size()) - 1;
	_Strings2ID.insert(kfPair<kfstr, int32>(
		namestr,nameid
		));
	return nameid;
}

const kfstr& KFDNameStrings::GetNameStr(int32 nameid)
{
	if (nameid < _ID2Strings.size())
	{
		return _ID2Strings[nameid];
	}
	return NULLSTR;
}


KF::KFDNameStrings& KFDName::NameStrings()
{
	static KFDNameStrings _NameStrings;
	return _NameStrings;
}

NS_KF_END





